consensus/fork_choice/src/fork_choice.rs

```diff
@@ -679,29 +705,110 @@ where
             self.fc_store.set_proposer_boost_root(block_root);
         }
 
-        // Update justified checkpoint.
-        if state.current_justified_checkpoint().epoch > self.fc_store.justified_checkpoint().epoch {
-            if state.current_justified_checkpoint().epoch
-                > self.fc_store.best_justified_checkpoint().epoch
+        let update_justified_checkpoint_slots = UpdateJustifiedCheckpointSlots::OnBlock {
+            state_slot: state.slot(),
+            current_slot,
+        };
+
+        // Update store with checkpoints if necessary
+        self.update_checkpoints(
+            state.current_justified_checkpoint(),
+            state.finalized_checkpoint(),
+            update_justified_checkpoint_slots,
+            spec,
+        )?;
+
+        // Update unrealized justified/finalized checkpoints.
+        let (unrealized_justified_checkpoint, unrealized_finalized_checkpoint) = if count_unrealized
+            .is_true()
+        {
+            let block_epoch = block.slot().epoch(E::slots_per_epoch());
+
+            // If the parent checkpoints are already at the same epoch as the block being imported,
+            // it's impossible for the unrealized checkpoints to differ from the parent's. This
+            // holds true because:
+            //
+            // 1. A child block cannot have lower FFG checkpoints than its parent.
+            // 2. A block in epoch `N` cannot contain attestations which would justify an epoch higher than `N`.
+            // 3. A block in epoch `N` cannot contain attestations which would finalize an epoch higher than `N - 1`.
+            //
+            // This is an optimization. It should reduce the amount of times we run
+            // `process_justification_and_finalization` by approximately 1/3rd when the chain is
+            // performing optimally.
+            let parent_checkpoints = parent_block
+                .unrealized_justified_checkpoint
+                .zip(parent_block.unrealized_finalized_checkpoint)
+                .filter(|(parent_justified, parent_finalized)| {
+                    parent_justified.epoch == block_epoch
+                        && parent_finalized.epoch + 1 >= block_epoch
```

Comment:
`parent_finalized.epoch + 1 > block_epoch` isn't actually possible right? If not I think it'd be better to update this to avoid confusion
```suggestion
                        && parent_finalized.epoch + 1 == block_epoch
```

Reply:
So I think the `>=` is useful during the genesis epoch. The `>=` means that we'll compute the unrealized values once and then duplicate them for the rest of the epoch, which is valid since nothing can be justified or finalized during the genesis epoch.

```rust
// During genesis
assert!(parent_justified.epoch == 0);
assert!(parent_finalized.epoch == 0);

assert!(block_epoch == 0);
let lhs = parent_justified.epoch == block_epoch;
let rhs = parent_finalized + 1 >= block_epoch;
assert!(lhs && rhs);
```

I don't think it would be an actual problem to have `==`, but having `>=` feels most correct to me.

Reply:
A genesis edge case -- classic! All good with me thn